command.go

```diff
@@ -1449,13 +1454,101 @@ func (c *Command) UseLine() string {
 	} else {
 		useline = use
 	}
+	return useline + c.uselineFlags(useline)
+}
+
+func (c *Command) uselineFlags(useline string) string {
 	if c.DisableFlagsInUseLine {
-		return useline
+		return ""
 	}
-	if c.HasAvailableFlags() && !strings.Contains(useline, "[flags]") {
-		useline += " [flags]"
+	if !c.HasAvailableFlags() {
+		return ""
 	}
-	return useline
+	if strings.Contains(useline, "[flags]") {
+		return ""
+	}
+	if c.DisableVerboseFlagsInUseLine {
+		return " [flags]"
+	}
+
+	included := map[*flag.Flag]struct{}{}
+	flagsLine := ""
+
+	c.flags.VisitAll(func(f *flag.Flag) {
+		if _, ok := included[f]; ok || f.Hidden {
+			return
+		}
+		included[f] = struct{}{}
+
+		rag := flagsFromAnnotation(c, f, requiredAsGroup)
+		me := flagsFromAnnotation(c, f, mutuallyExclusive)
+		or := flagsFromAnnotation(c, f, oneRequired)
+
+		if len(rag) > 0 {
+			gr := []string{}
+			for _, fl := range rag {
+				included[fl] = struct{}{}
+				gr = append(gr, shortUsage(fl))
+			}
+			flagsLine += " [" + strings.Join(gr, " ") + "]"
+		} else if len(me) > 0 {
+			gr := []string{}
+			for _, fl := range me {
+				included[fl] = struct{}{}
+				gr = append(gr, shortUsage(fl))
+			}
+			if sameFlags(me, or) {
+				flagsLine += " {" + strings.Join(gr, " | ") + "}"
+			} else {
+				flagsLine += " [" + strings.Join(gr, " | ") + "]"
+			}
+		} else if req, found := f.Annotations[BashCompOneRequiredFlag]; found && req[0] == "true" {
+			flagsLine += " " + shortUsage(f)
+		} else {
+			flagsLine += " [" + shortUsage(f) + "]"
+		}
+	})
```

Comment:
The code to list the short usage strings is repeated twice, making this function even more complicated and hard to follow. We have here many very short variable names (arg, me, or, gr, fl) and complex string formatting. It is pain to follow this code.

We can make it little better by extracting few helpers. I'm not sure about the names, they can be better.

```Go
func groupShortUsage(flags map[string]*flag.Flag, included map[*flag.Flag]struct{}) []string {
    gr := []string{}
    for _, fl := range flags {
        included[fl] = struct{}{}
        gr = append(gr, shortUsage(fl))
    }
    return gr
}

func isOneRequired(f *flag.Flag) bool {
    req, found := f.Annotations[BashCompOneRequiredFlag]
    return found && req[0] == "true"    
}
```

Now we have only 3 very short variables names which is easy to grasp, the rest is the obvious formatting.

```Go
		rag := flagsFromAnnotation(c, f, requiredAsGroup)
		me := flagsFromAnnotation(c, f, mutuallyExclusive)
		or := flagsFromAnnotation(c, f, oneRequired)

		if len(rag) > 0 {
			usages := groupShortUsage(rag, included)
			flagsLine += " [" + strings.Join(usages, " ") + "]"
		} else if len(me) > 0 {
			usages := groupShortUsage(me, included)
			if sameFlags(me, or) {
				flagsLine += " {" + strings.Join(usages, " | ") + "}"
			} else {
				flagsLine += " [" + strings.Join(usages, " | ") + "]"
			}
		} else if isOneRequired(f) {
			flagsLine += " " + shortUsage(f)
		} else {
			flagsLine += " [" + shortUsage(f) + "]"
		}
```